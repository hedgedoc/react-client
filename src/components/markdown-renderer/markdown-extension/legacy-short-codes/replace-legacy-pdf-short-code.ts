/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'
import type { RegexOptions } from '../../../../external-types/markdown-it-regex/interface'

const finalRegex = /^{%pdf ([^ ]*) ?%}$/

export const legacyPdfShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-pdf-short-code',
    regex: finalRegex,
    replace: (match) => `<a href="${match}">${match}</a>`
  } as RegexOptions)
}
