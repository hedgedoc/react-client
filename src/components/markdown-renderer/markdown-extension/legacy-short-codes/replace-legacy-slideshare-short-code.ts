/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'
import type { RegexOptions } from '../../../../external-types/markdown-it-regex/interface'

const finalRegex = /^{%slideshare (\w+\/[\w-]+) ?%}$/

/**
 * Configure the given {@link MarkdownIt} to render legacy hedgedoc 1 slideshare shortcodes as HTML links.
 *
 * @param markdownIt The {@link MarkdownIt} to configure
 */
export const legacySlideshareShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-slideshare-short-code',
    regex: finalRegex,
    replace: (match) => `<a href='https://www.slideshare.net/${match}'>https://www.slideshare.net/${match}</a>`
  } as RegexOptions)
}
