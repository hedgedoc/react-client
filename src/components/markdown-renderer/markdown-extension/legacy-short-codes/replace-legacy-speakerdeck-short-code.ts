/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'
import type { RegexOptions } from '../../../../external-types/markdown-it-regex/interface'

const finalRegex = /^{%speakerdeck (\w+\/[\w-]+) ?%}$/

export const legacySpeakerdeckShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-speakerdeck-short-code',
    regex: finalRegex,
    replace: (match) => `<a href="https://speakerdeck.com/${match}">https://speakerdeck.com/${match}</a>`
  } as RegexOptions)
}
