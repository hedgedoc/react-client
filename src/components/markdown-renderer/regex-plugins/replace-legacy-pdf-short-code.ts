/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'

const finalRegex = /^{%pdf (.*) ?%}$/

export const legacyPdfShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-pdf-short-code',
    regex: finalRegex,
    replace: (match: string) => {
      return `<a target="_blank" rel="noopener noreferrer" href="${match}">${match}</a>`
    }
  })
}
