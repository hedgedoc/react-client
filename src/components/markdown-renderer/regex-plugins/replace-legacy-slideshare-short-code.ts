/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'

const finalRegex = /^{%slideshare (\w+\/[\w-]+) ?%}$/

export const legacySlideshareShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-slideshare-short-code',
    regex: finalRegex,
    replace: (match: string) => {
      return `<a target="_blank" rel="noopener noreferrer" href="https://www.slideshare.net/${match}">https://www.slideshare.net/${match}</a>`
    }
  })
}
