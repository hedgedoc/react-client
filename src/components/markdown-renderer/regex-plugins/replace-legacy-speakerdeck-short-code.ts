/*
 * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import markdownItRegex from 'markdown-it-regex'
import type MarkdownIt from 'markdown-it/lib'

const finalRegex = /^{%speakerdeck (\w+\/[\w-]+) ?%}$/

export const legacySpeakerdeckShortCode: MarkdownIt.PluginSimple = (markdownIt) => {
  markdownItRegex(markdownIt, {
    name: 'legacy-speakerdeck-short-code',
    regex: finalRegex,
    replace: (match: string) => {
      return `<a target="_blank" rel="noopener noreferrer" href="https://speakerdeck.com/${match}">https://speakerdeck.com/${match}</a>`
    }
  })
}
