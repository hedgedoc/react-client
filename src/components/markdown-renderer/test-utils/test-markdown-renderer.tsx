/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import React, { useMemo } from 'react'
import { useConvertMarkdownToReactDom } from '../hooks/use-convert-markdown-to-react-dom'
import type { MarkdownExtension } from '../markdown-extension/markdown-extension'
import { StoreProvider } from '../../../redux/store-provider'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownExtension[]
}

export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  const dom = useConvertMarkdownToReactDom(lines, extensions, true, false)

  return <StoreProvider>{dom}</StoreProvider>
}
