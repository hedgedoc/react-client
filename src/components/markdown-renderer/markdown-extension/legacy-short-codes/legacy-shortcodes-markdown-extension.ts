/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { MarkdownExtension } from '../markdown-extension'
import type MarkdownIt from 'markdown-it'
import { legacyPdfShortCode } from './replace-legacy-pdf-short-code'
import { legacySlideshareShortCode } from './replace-legacy-slideshare-short-code'
import { legacySpeakerdeckShortCode } from './replace-legacy-speakerdeck-short-code'

/**
 * Adds support for legacy shortcodes (pdf, slideshare and speakerdeck) by replacing them with anchor elements.
 */
export class LegacyShortcodesMarkdownExtension extends MarkdownExtension {
  public configureMarkdownIt(markdownIt: MarkdownIt): void {
    legacyPdfShortCode(markdownIt)
    legacySlideshareShortCode(markdownIt)
    legacySpeakerdeckShortCode(markdownIt)
  }
}
