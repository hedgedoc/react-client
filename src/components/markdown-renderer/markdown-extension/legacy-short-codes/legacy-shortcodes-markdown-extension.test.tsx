/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as replaceLegacyPdfShortCodeModule from './replace-legacy-pdf-short-code'
import * as replaceLegacySlideshareShortCodeModule from './replace-legacy-slideshare-short-code'
import * as replaceLegacySpeakerdeckShortCodeModule from './replace-legacy-speakerdeck-short-code'
import { render } from '@testing-library/react'
import { TestMarkdownRenderer } from '../../test-utils/test-markdown-renderer'
import React from 'react'
import { LegacyShortcodesMarkdownExtension } from './legacy-shortcodes-markdown-extension'
import MarkdownIt from 'markdown-it'

describe('Legacy shortcodes markdown extension', () => {
  const replaceLegacyPdfShortCodeMarkdownItExtensionMock = jest.spyOn(
    replaceLegacyPdfShortCodeModule,
    'legacyPdfShortCode'
  )
  const replaceLegacySlideshareShortCodeMarkdownItExtensionMock = jest.spyOn(
    replaceLegacySlideshareShortCodeModule,
    'legacySlideshareShortCode'
  )
  const replaceLegacySpeakerdeckShortCodeMarkdownItExtensionMock = jest.spyOn(
    replaceLegacySpeakerdeckShortCodeModule,
    'legacySpeakerdeckShortCode'
  )

  it('renders correctly', () => {
    const view = render(
      <TestMarkdownRenderer
        extensions={[new LegacyShortcodesMarkdownExtension()]}
        content={
          '{%pdf https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf %}\n{%slideshare example/123456789 %}\n{%speakerdeck example/123456789 %}'
        }
      />
    )
    expect(replaceLegacyPdfShortCodeMarkdownItExtensionMock).toHaveBeenCalledWith(expect.any(MarkdownIt))
    expect(replaceLegacySlideshareShortCodeMarkdownItExtensionMock).toHaveBeenCalledWith(expect.any(MarkdownIt))
    expect(replaceLegacySpeakerdeckShortCodeMarkdownItExtensionMock).toHaveBeenCalledWith(expect.any(MarkdownIt))
    expect(view.container).toMatchSnapshot()
  })
})
