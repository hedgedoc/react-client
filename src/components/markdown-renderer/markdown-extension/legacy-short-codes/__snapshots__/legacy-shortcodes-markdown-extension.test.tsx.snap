// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legacy shortcodes markdown extension renders correctly 1`] = `
<div>
  <p>
    <a
      href="https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf"
    >
      https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf
    </a>
    <br />
    

    <a
      href="https://www.slideshare.net/example/123456789"
    >
      https://www.slideshare.net/example/123456789
    </a>
    <br />
    

    <a
      href="https://speakerdeck.com/example/123456789"
    >
      https://speakerdeck.com/example/123456789
    </a>
  </p>
  

</div>
`;
